File: lib/flowpace/store.ts
import { promises as fs } from 'fs';
import path from 'path';
import type { Routine, CompletedRoutine, AppSettings } from './types';

const DATA_DIR = process.env.VERCEL
  ? '/tmp/flowpace'
  : path.join(process.cwd(), 'data', 'flowpace');

const ROUTINES_FILE = path.join(DATA_DIR, 'routines.json');
const COMPLETED_FILE = path.join(DATA_DIR, 'completed.json');
const SETTINGS_FILE = path.join(DATA_DIR, 'settings.json');

const DEFAULT_SETTINGS: AppSettings = {
  is_pro: false,
  is_audio_enabled: true,
  is_voice_enabled: false,
  volume: 0.7,
  selected_sound_pack: 'default',
  background_color: 'blue',
};

let memoryRoutines: Routine[] | null = null;
let memoryCompleted: CompletedRoutine[] | null = null;
let memorySettings: AppSettings | null = null;

async function ensureDir() {
  try {
    await fs.mkdir(DATA_DIR, { recursive: true });
  } catch {
    // ignore
  }
}

async function readJson<T>(file: string, fallback: T): Promise<T> {
  try {
    const text = await fs.readFile(file, 'utf-8');
    return JSON.parse(text) as T;
  } catch {
    return fallback;
  }
}

async function writeJson(file: string, data: unknown) {
  try {
    await ensureDir();
    await fs.writeFile(file, JSON.stringify(data, null, 2));
  } catch {
    // ignore (read-only fs)
  }
}

export async function getRoutines(): Promise<Routine[]> {
  if (memoryRoutines) return memoryRoutines;
  memoryRoutines = await readJson<Routine[]>(ROUTINES_FILE, []);
  return memoryRoutines;
}

export async function getRoutine(routineId: string): Promise<Routine | undefined> {
  const routines = await getRoutines();
  return routines.find((r) => r.id === routineId);
}

export async function addRoutine(routine: Routine): Promise<Routine> {
  const routines = await getRoutines();
  routines.push(routine);
  memoryRoutines = routines;
  await writeJson(ROUTINES_FILE, routines);
  return routine;
}

export async function updateRoutine(routine: Routine): Promise<Routine | null> {
  const routines = await getRoutines();
  const idx = routines.findIndex((r) => r.id === routine.id);
  if (idx === -1) return null;
  routines[idx] = routine;
  memoryRoutines = routines;
  await writeJson(ROUTINES_FILE, routines);
  return routine;
}

export async function deleteRoutine(routineId: string): Promise<boolean> {
  const routines = await getRoutines();
  const before = routines.length;
  const filtered = routines.filter((r) => r.id !== routineId);
  if (filtered.length === before) return false;
  memoryRoutines = filtered;
  await writeJson(ROUTINES_FILE, filtered);
  return true;
}

export async function reorderRoutines(orderedIds: string[]) {
  const routines = await getRoutines();
  const idMap = new Map(routines.map((r) => [r.id, r]));
  const reordered = orderedIds
    .map((id) => idMap.get(id))
    .filter((r): r is Routine => !!r);
  memoryRoutines = reordered;
  await writeJson(ROUTINES_FILE, reordered);
}

export async function getCompletions(): Promise<CompletedRoutine[]> {
  if (memoryCompleted) return memoryCompleted;
  memoryCompleted = await readJson<CompletedRoutine[]>(COMPLETED_FILE, []);
  return memoryCompleted;
}

export async function recordCompletion(completion: CompletedRoutine) {
  const completions = await getCompletions();
  completions.push(completion);
  memoryCompleted = completions;
  await writeJson(COMPLETED_FILE, completions);
}

export async function getSettings(): Promise<AppSettings> {
  if (memorySettings) return memorySettings;
  memorySettings = await readJson<AppSettings>(SETTINGS_FILE, DEFAULT_SETTINGS);
  return memorySettings;
}

export async function updateSettings(settings: AppSettings) {
  memorySettings = settings;
  await writeJson(SETTINGS_FILE, settings);
}

File: flowpace/api/lib/store.py
from __future__ import annotations

import json
import os
from datetime import datetime
from pathlib import Path
from typing import List, Optional

from .models import CompletedRoutine, Routine, AppSettings

# In-memory store (shared across requests in the same serverless instance)
_mem = {
    "routines": [],
    "completed": [],
    "settings": AppSettings(),
}

# Optional JSON persistence for local dev (vercel dev has a writable fs)
_data_dir = Path(__file__).parent.parent.parent / "data"
_routines_file = _data_dir / "routines.json"
_completed_file = _data_dir / "completed.json"
_settings_file = _data_dir / "settings.json"


def _to_json(obj):
    if isinstance(obj, datetime):
        return obj.isoformat()
    raise TypeError


def _from_json(dct):
    for k, v in dct.items():
        if k == "completed_at" and isinstance(v, str):
            try:
                dct[k] = datetime.fromisoformat(v)
            except ValueError:
                pass
    return dct


def _load_json():
    if _routines_file.exists():
        try:
            data = json.loads(_routines_file.read_text())
            _mem["routines"] = [Routine(**r) for r in data]
        except Exception:
            pass
    if _completed_file.exists():
        try:
            data = json.loads(_completed_file.read_text(), object_hook=_from_json)
            _mem["completed"] = [CompletedRoutine(**c) for c in data]
        except Exception:
            pass
    if _settings_file.exists():
        try:
            data = json.loads(_settings_file.read_text())
            _mem["settings"] = AppSettings(**data)
        except Exception:
            pass


def _save_json():
    try:
        _data_dir.mkdir(exist_ok=True)
        _routines_file.write_text(json.dumps([r.model_dump() for r in _mem["routines"]], default=_to_json))
        _completed_file.write_text(json.dumps([c.model_dump() for c in _mem["completed"]], default=_to_json))
        _settings_file.write_text(json.dumps(_mem["settings"].model_dump()))
    except Exception:
        pass


def get_routines() -> List[Routine]:
    _load_json()
    return _mem["routines"]


def get_routine(routine_id: str) -> Optional[Routine]:
    _load_json()
    return next((r for r in _mem["routines"] if r.id == routine_id), None)


def add_routine(routine: Routine) -> Routine:
    _load_json()
    _mem["routines"].append(routine)
    _save_json()
    return routine


def update_routine(routine: Routine) -> Optional[Routine]:
    _load_json()
    for i, r in enumerate(_mem["routines"]):
        if r.id == routine.id:
            _mem["routines"][i] = routine
            _save_json()
            return routine
    return None


def delete_routine(routine_id: str) -> bool:
    _load_json()
    before = len(_mem["routines"])
    _mem["routines"] = [r for r in _mem["routines"] if r.id != routine_id]
    if len(_mem["routines"]) != before:
        _save_json()
        return True
    return False


def reorder_routines(ordered_ids: List[str]):
    _load_json()
    id_map = {r.id: r for r in _mem["routines"]}
    _mem["routines"] = [id_map[rid] for rid in ordered_ids if rid in id_map]
    _save_json()


def record_completion(completion: CompletedRoutine):
    _load_json()
    _mem["completed"].append(completion)
    _save_json()


def get_completions() -> List[CompletedRoutine]:
    _load_json()
    return _mem["completed"]


def get_settings() -> AppSettings:
    _load_json()
    return _mem["settings"]


def update_settings(settings: AppSettings):
    _mem["settings"] = settings
    _save_json()
